### PR TITLE
Throw exception when server heartbeat missed

### PR DESCRIPTION
**Pull request by fuhry**
There is currently no handling for a missed heartbeat event from the server. This pull request uses the local heartbeat timer, `$this->heartbeat` and throws an `AMQPRuntimeException` if `$this->heartbeat * 2` seconds pass without receiving any frame from the server.

I still don't have a great mental picture of the timeout handling and blocking behavior in PhpAmqpLib. If the user specifies a timeout value to the `wait` method that is greater than `$this->heartbeat`, or does not specify a timeout at all thus causing the IO layer to wait indefinitely for traffic, we will still be at the mercy of the host OS to close the dead socket. Is this behavior correct? Or should we continue waiting until the user's requested timeout, but make the internal read timeout sooner and throw the exception if a server heartbeat is missed before the user-specified timeout is reached?

**PLEASE REVIEW:** when `$this->heartbeat` is set, does this also request the AMQP server to send heartbeat frames every `$this->heartbeat` seconds? Or is that configurable only on the server? If the latter case is true, then this PR should be rejected, and new methods and parameters need to be added to instruct the client to enforce a specified server heartbeat interval.

### CODE REVIEW

**Comment by postalservice14** [comment]
I'm trying to figure out how to test this (even manually).  Are you running into issues where the server heartbeat is not being sent?

To answer your question about the client setting the heartbeat time, yes, if you set the heartbeat time to say 10 seconds, it will send a frame to the server setting the requested heartbeat time and the server will send a heartbeat every 10 seconds.

**Comment by fuhry** [comment]
@postalservice14 My testing process was to add an `iptables -A OUTPUT -j DROP` rule to the AMQP server to simulate a connection loss. Another approach, if your test server is in a VM, is to have the hypervisor unplug the virtual network cable.

Obviously, something like this is not appropriate for unit testing or even a more extensive integration test suite. I'm not immediately aware of any tool kits that allow PHP network applications to be subjected to adverse network conditions, so this might have to be manually tested for now.

**Comment by michaelklishin** [comment]
Force closing connections using HTTP API or `rabbitmqctl` is what several client test suites use (Ruby clients, Java, .NET, Clojure, …).

**Comment by michaelklishin** [comment]
```
if you set the heartbeat time to say 10 seconds, it will send a frame to the server setting the requested heartbeat time and the server will send a heartbeat every 10 seconds
```

@postalservice14 sadly different clients treat this setting slightly differently. In case this is an *interval* then the above is correct. In case it is a *timeout*, a heartbeat frame should be sent at half the interval of the timeout (because 2 or more missed heartbeats should be considered a timeout).

It is more correct and also safer to do the latter and explain in the docs that heartbeat timeout values < 5 seconds are likely to result in false positives, so should be avoided.

**Comment by postalservice14** [comment]
Sorry, I haven't looked at this in a while.  But I think it looks good to me.  @michaelklishin and @fuhry , you guys seem pretty knowledgable on the subject.  Does the proposed PR make sense to merge in?

File: PhpAmqpLib/Connection/AbstractConnection.php [hunk 1]
```diff
@@ -78,6 +78,9 @@ class AbstractConnection extends AbstractChannel
     /** @var int */
     protected $heartbeat;
 
+    /** @var float */
+    protected $last_frame;
```
Comment (postalservice14) [inline PhpAmqpLib/Connection/AbstractConnection.php]: The only "code-style" comment I have is I want to start moving towards PSR-2 as a coding standard.  Can you make this variable `$lastFrame` instead?

Comment (prolic) [inline PhpAmqpLib/Connection/AbstractConnection.php]: That kind of change should come for the whole project and not partially, to be consistent at least.

Comment (fuhry) [inline PhpAmqpLib/Connection/AbstractConnection.php]: Agreed. Using inconsistent naming within the same file, or in this case, on the same line of code is more likely to lead to mistakes than a single large refactor. Consistency > partial standards conformance.

**Comment by fuhry** [comment]
@postalservice14 We are no longer using AMQP for the product in question. However, during the time we were testing this patch it did generally improve reliability quite a bit. It's been a good 4-5 months since we have tested anything AMQP related though.

**Comment by michaelklishin** [comment]
@nubeiro @hairyhum WDYT?

**Comment by andrefigueira** [comment]
I'm actually coming across this a lot in production where we're having connections being dropped off and heartbeats being missed, and no exception is being thrown, so my consumers are still running even though they will never receive a message again, it's only happening with specific versions of rabbit, but it's definitely down to this.

**Comment by lukebakken** [comment]
@michaelklishin this seems reasonable to me, thought I would prefer a configurable value [here](https://github.com/php-amqplib/php-amqplib/pull/354/files#diff-46e92955bdff3b3f843a50f6344bd32fR556).

**Comment by michaelklishin** [comment]
@lukebakken what specifically would you like to see configurable?

**Comment by lukebakken** [comment]
@michaelklishin it is this multiplier - `(at threshold * 2)` ... but that seems standard. Thanks for reviewing

**Pull request merged**